runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+    if start_row ~= end_row then
+      row_text[start_row] = range.collapsedText
+      for row = start_row, end_row do
+        local level = row_level[row] or { 0 }
+        level[1] = level[1] + 1
+        row_level[row] = level
+      end
+      row_level[start_row][2] = '>'
+      row_level[end_row][2] = '<'
+    end
+  end
+
+  bufstate.row_level = row_level
+  bufstate.row_text = row_text
+  bufstate.kind_ranges = kind_ranges
+  bufstate.version = ctx.version
+
+  if api.nvim_get_mode().mode:match('^i') then
+    -- `foldUpdate()` is guarded in insert mode.
+    schedule_foldupdate(bufnr)
+  else
+    foldupdate(bufnr)
+  end
+end
+
+local augroup_refresh = vim.api.nvim_create_augroup('vim_lsp_folding_range/refresh', {})
+
+---@param bufnr integer
+---@param client_id integer The ID of the client whose server supports `foldingRange`.
+---@return integer? request_id
+local function refresh(bufnr, client_id)
+  ---@type lsp.FoldingRangeParams
+  local params = { textDocument = util.make_text_document_params(bufnr) }
+  local client = assert(vim.lsp.get_client_by_id(client_id))
+  local _, request_id = client:request(ms.textDocument_foldingRange, params, nil, bufnr)
+  return request_id
+end
+
+--- Setup event hooks for folding range.
+---@param bufnr integer
+---@return vim.lsp.folding_range.BufState
+local function setup(bufnr)
+  ---@type vim.lsp.folding_range.BufState
+  local bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+  bufstates[bufnr] = bufstate
+
+  api.nvim_buf_attach(bufnr, false, {
+    on_detach = function()
+      bufstates[bufnr] = nil
+    end,
+    on_reload = function()
+      bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+      bufstates[bufnr] = bufstate
+      for _, client in
+        ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+      do
+        refresh(bufnr, client.id)
+      end
+    end,
+    --- When modifying the buffer, new foldinfo is scheduled to be updated,
+    --- synchronize the changed rows with their previous foldlevels as possible,
+    --- before new foldlevels is updated.
+    ---@param start_row integer
+    ---@param old_row integer
+    ---@param new_row integer
+    on_bytes = function(_, _, _, start_row, _, _, old_row, _, _, new_row, _, _)
+      local row = new_row - old_row
+      local row_level = bufstate.row_level
+      if vim.tbl_isempty(row_level) then
+        return
+      end
+      if row > 0 then
+        vim._list_insert(bufstate.row_level, start_row, start_row + math.abs(row) - 1, { -1 })
+      elseif row < 0 then
+        vim._list_remove(bufstate.row_level, start_row, start_row + math.abs(row) - 1)
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspDetach', {
+    group = augroup_refresh,
+    buffer = bufnr,
+    callback = function()
+      local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange })
+      if vim.tbl_isempty(clients) then
```

Comment:
```suggestion
      if next(clients) then
```

Reply:
Because it is standard Lua. No need to use a redundant wrapper.

Reply:
Looking back, I think `#client == 0` should be sufficient, and its meaning should be clearer.